web/packages/teleport/src/ironrdp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+                }
+                ActiveStageOutput::ResponseFrame(frame) => {
+                    // Send the response frame back to the server.
+                    let frame = Uint8Array::from(frame.as_slice()); // todo(isaiah): this is a copy
+                    let _ = respond_cb.call1(cb_context, &frame.buffer())?;
+                }
+                ActiveStageOutput::Terminate => {
+                    return Err(JsValue::from_str("Terminate should never be returned"));
+                }
+                _ => {
+                    debug!("Unhandled ActiveStageOutput: {:?}", output);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn apply_image_to_canvas(
+        &self,
+        image_data: Vec<u8>,
+        image_location: InclusiveRectangle,
+        cb_context: &JsValue,
+        callback: &js_sys::Function,
+    ) -> Result<(), JsValue> {
+        let top = image_location.top;
+        let left = image_location.left;
+
+        let image_data = create_image_data_from_image_and_region(&image_data, image_location)?;
+        let bitmap_frame = BitmapFrame {
+            top,
+            left,
+            image_data,
+        };
+
+        let bitmap_frame = &JsValue::from(bitmap_frame);
+
+        // TODO(isaiah): return this?
+        let _ret = callback.call1(cb_context, bitmap_frame)?;
+        Ok(())
+    }
+}
+
+pub fn extract_partial_image(
+    image: &DecodedImage,
+    region: InclusiveRectangle,
+) -> (InclusiveRectangle, Vec<u8>) {
+    // PERF: needs actual benchmark to find a better heuristic
+    if region.height() > 64 || region.width() > 512 {
+        extract_whole_rows(image, region)
+    } else {
+        extract_smallest_rectangle(image, region)
+    }
+}
+
+// Faster for low-height and smaller images
+fn extract_smallest_rectangle(
+    image: &DecodedImage,
+    region: InclusiveRectangle,
+) -> (InclusiveRectangle, Vec<u8>) {
+    let pixel_size = usize::from(image.pixel_format().bytes_per_pixel());
+
+    let image_width = usize::try_from(image.width()).unwrap();
+    let image_stride = image_width * pixel_size;
+
+    let region_top = usize::from(region.top);
+    let region_left = usize::from(region.left);
+    let region_width = usize::from(region.width());
+    let region_height = usize::from(region.height());
+    let region_stride = region_width * pixel_size;
+
+    let dst_buf_size = region_width * region_height * pixel_size;
+    let mut dst = vec![0; dst_buf_size];
+
+    let src = image.data();
+
+    for row in 0..region_height {
+        let src_begin = image_stride * (region_top + row) + region_left * pixel_size;
+        let src_end = src_begin + region_stride;
+        let src_slice = &src[src_begin..src_end];
+
+        let target_begin = region_stride * row;
+        let target_end = target_begin + region_stride;
+        let target_slice = &mut dst[target_begin..target_end];
```

Comment:
This seems correct but the price of being incorrect is a panic here, so this makes me a bit uneasy. Is there some more reasonable way of writing this?

Reply:
For instance, what if `region` is not contained in the image bounds?